src/python_pachyderm/mixin/util.py

```diff
@@ -1,18 +1,35 @@
+import re
+
 from python_pachyderm.proto.v2.pfs import pfs_pb2 as pfs_proto
 
+# copied from pachyderm/pachyderm
+uuid_re = re.compile(r"[0-9a-f]{12}4[0-9a-f]{19}")
+valid_branch_re = re.compile(r"^[a-zA-Z0-9_-]+$")
+
 
-def commit_from(src, allow_just_repo=False):
+def commit_from(src):
+    # TODO: refactor this overloading of repo/commit in a single string
     if src is None:
         return None
-
     if isinstance(src, pfs_proto.Commit):
         return src
-    elif isinstance(src, (tuple, list)) and len(src) == 2:
-        return pfs_proto.Commit(repo=pfs_proto.Repo(name=src[0]), id=src[1])
+
+    repo, branch, commit = None, None, None
+    if isinstance(src, (tuple, list)) and len(src) == 2:
+        repo, branch_or_commit = src
     elif isinstance(src, str):
-        repo_name, commit_id = src.split("/", 1)
-        return pfs_proto.Commit(repo=pfs_proto.Repo(name=repo_name), id=commit_id)
+        repo, branch_or_commit = src.split("/", 1)
+    if uuid_re.match(branch_or_commit) or not valid_branch_re.match(branch_or_commit):
+        commit = branch_or_commit
```

Comment:
It seems like, if the code takes this branch, then `branch` is `None`, which I think is invalid (e.g. for `ListFile`—I think if `pfs_proto.Commit.id` is `None`, then the most recent commit is assumed, but if `pfs_proto.Commit.branch.name` is `None`, then that's an error)

Reply:
Yes this was a bit confusing to me at first. However, I think the server can take in an empty branch name. At least the way we currently parse inputs to `pachctl` suggests this.

https://github.com/pachyderm/pachyderm/blob/3b69deea92a21ba5877b8792b99ba5acfd1258e0/src/internal/cmdutil/cobra.go#L156-L159

Reply:
Interesting. I may ping Marc to ask what the semantics of this is—I'm not sure what pachyderm does with a commit but no branch (especially since the test suite needed to start setting `"master"` everywhere)